File: sdk/python/gestalt/public/rest_transport.py
"""httpx-based REST transports for the public Gestalt API (/api/v2).

``RestUnaryTransport`` is sync (backed by ``httpx.Client``);
``AsyncRestUnaryTransport`` is async (backed by ``httpx.AsyncClient``). Both
share request building and response decoding via ``rest_request``.
"""

from __future__ import annotations

from collections.abc import Mapping
from typing import Any, Protocol, TypeVar, cast

import httpx
from google.protobuf.message import Message

from .auth import AuthProvider
from .generated.metadata import Method
from .rest_request import (
    build_rest_request,
    decode_rest_response,
    map_httpx_error,
)

ResponseT = TypeVar("ResponseT", bound=Message)


class HttpClient(Protocol):
    def request(
        self,
        method: str,
        url: str,
        *,
        headers: Mapping[str, str] | None = None,
        params: str | None = None,
        json: dict[str, Any] | None = None,
        timeout: float | None = None,
    ) -> httpx.Response: ...


class AsyncHttpClient(Protocol):
    async def request(
        self,
        method: str,
        url: str,
        *,
        headers: Mapping[str, str] | None = None,
        params: str | None = None,
        json: dict[str, Any] | None = None,
        timeout: float | None = None,
    ) -> httpx.Response: ...


class RestUnaryTransport:
    """Protobuf-JSON REST transport implementing the generated UnaryTransport protocol."""

    def __init__(
        self,
        base_url: str,
        auth: AuthProvider,
        *,
        client: HttpClient | None = None,
    ) -> None:
        self._base_url = base_url.rstrip("/")
        self._auth = auth
        self._client = client
        self._owns_client = client is None
        self._owned_client: httpx.Client | None = None
        if self._owns_client:
            self._owned_client = httpx.Client(
                verify=True,
                transport=httpx.HTTPTransport(retries=0),
            )

    def unary(
        self,
        method: Method,
        request: Message,
        response_type: type[ResponseT],
    ) -> ResponseT:
        prepared = build_rest_request(method, request, self._base_url, self._auth)
        client = self._client or self._owned_client
        if client is None:
            raise RuntimeError("REST transport client is not available")

        try:
            response = client.request(
                prepared.verb,
                prepared.url,
                headers=prepared.headers,
                params=prepared.params,
                json=prepared.json_body,
            )
        except (httpx.TimeoutException, httpx.HTTPError) as err:
            raise map_httpx_error(err) from err

        header_map = {key: value for key, value in response.headers.items()}
        return cast(
            ResponseT,
            decode_rest_response(
                response_type, response.status_code, response.content, header_map
            ),
        )

    def close(self) -> None:
        if self._owns_client and self._owned_client is not None:
            self._owned_client.close()
            self._owned_client = None


class AsyncRestUnaryTransport:
    """Async protobuf-JSON REST transport implementing the generated AsyncUnaryTransport protocol."""

    def __init__(
        self,
        base_url: str,
        auth: AuthProvider,
        *,
        client: AsyncHttpClient | None = None,
    ) -> None:
        self._base_url = base_url.rstrip("/")
        self._auth = auth
        self._client = client
        self._owns_client = client is None
        self._owned_client: httpx.AsyncClient | None = None
        if self._owns_client:
            self._owned_client = httpx.AsyncClient(
                verify=True,
                transport=httpx.AsyncHTTPTransport(retries=0),
            )

    async def unary(
        self,
        method: Method,
        request: Message,
        response_type: type[ResponseT],
    ) -> ResponseT:
        prepared = build_rest_request(method, request, self._base_url, self._auth)
        client = self._client or self._owned_client
        if client is None:
            raise RuntimeError("async REST transport client is not available")

        try:
            response = await client.request(
                prepared.verb,
                prepared.url,
                headers=prepared.headers,
                params=prepared.params,
                json=prepared.json_body,
            )
        except (httpx.TimeoutException, httpx.HTTPError) as err:
            raise map_httpx_error(err) from err

        header_map = {key: value for key, value in response.headers.items()}
        return cast(
            ResponseT,
            decode_rest_response(
                response_type, response.status_code, response.content, header_map
            ),
        )

    async def close(self) -> None:
        if self._owns_client and self._owned_client is not None:
            await self._owned_client.aclose()
            self._owned_client = None

File: sdk/rust/src/public/rest_transport.rs
//! reqwest-based REST transports for the public Gestalt API (/api/v2).
//!
//! [`RestTransport`] is async (backed by `reqwest`); [`SyncRestTransport`] is
//! sync (backed by `reqwest::blocking`). Both share request building and
//! response decoding via shared helpers in the rest_request module.

use std::sync::Arc;
use std::time::Duration;

use prost::Message;
use reqwest::header::HeaderMap;

use crate::public::auth::Auth;
use crate::public::generated::metadata::Method;
use crate::public::generated::rpc_support::GestaltError;
use crate::public::generated::unary_transport::{SyncUnaryTransport, UnaryTransport};
use crate::public::rest_request::{build_rest_request, decode_rest_response};
use crate::rpc_support::gestalt_error_code;

/// Async reqwest-based REST transport implementing [`UnaryTransport`].
#[derive(Clone)]
pub struct RestTransport {
    base_url: String,
    auth: Arc<dyn Auth>,
    client: reqwest::Client,
}

impl RestTransport {
    /// Creates a REST transport rooted at `base_url`.
    pub fn new(base_url: impl Into<String>, auth: Arc<dyn Auth>) -> Self {
        Self {
            base_url: base_url.into(),
            auth,
            client: reqwest::Client::builder()
                .use_rustls_tls()
                .build()
                .expect("reqwest client"),
        }
    }

    /// Applies a per-request timeout to the underlying HTTP client.
    pub fn with_timeout(mut self, timeout: Duration) -> Self {
        self.client = reqwest::Client::builder()
            .use_rustls_tls()
            .timeout(timeout)
            .build()
            .expect("reqwest client");
        self
    }
}

fn map_send_error(err: reqwest::Error) -> GestaltError {
    if err.is_timeout() {
        GestaltError::new(gestalt_error_code::DEADLINE_EXCEEDED, err.to_string())
    } else if err.is_request() {
        GestaltError::new(gestalt_error_code::INVALID_ARGUMENT, err.to_string())
    } else {
        GestaltError::new(gestalt_error_code::UNAVAILABLE, err.to_string())
    }
}

fn map_body_error(err: reqwest::Error) -> GestaltError {
    GestaltError::new(gestalt_error_code::UNAVAILABLE, err.to_string())
}

fn finalize_response<Resp>(
    status: u16,
    response_headers: &HeaderMap,
    body: &[u8],
    decode: crate::public::generated::metadata::DecodeResponseJson,
    response: &mut Resp,
) -> Result<(), GestaltError>
where
    Resp: Message + Default + Send,
{
    let response_bytes = decode_rest_response(status, response_headers, body, decode)?;
    *response = Resp::decode(response_bytes.as_slice())
        .map_err(|err| GestaltError::new(gestalt_error_code::INTERNAL, err.to_string()))?;
    Ok(())
}

impl UnaryTransport for RestTransport {
    fn unary<Req, Resp>(
        &self,
        method: &Method,
        request: &Req,
        response: &mut Resp,
    ) -> impl std::future::Future<Output = Result<(), GestaltError>> + Send
    where
        Req: Message + Send + Sync,
        Resp: Message + Default + Send,
    {
        let base_url = self.base_url.clone();
        let auth = Arc::clone(&self.auth);
        let client = self.client.clone();
        let request_bytes = request.encode_to_vec();
        let method = method.clone();

        async move {
            let decode = method.decode_response_json.ok_or_else(|| {
                GestaltError::new(
                    gestalt_error_code::INVALID_ARGUMENT,
                    format!("method {} has no REST decoder", method.full_method),
                )
            })?;
            let prepared = build_rest_request(&method, &request_bytes, &base_url, &auth)?;
            let mut builder = client
                .request(prepared.method, prepared.url)
                .headers(prepared.headers);
            if let Some(body) = prepared.body {
                builder = builder.json(&body);
            }
            let http_response = builder.send().await.map_err(map_send_error)?;
            let status = http_response.status();
            let response_headers = http_response.headers().clone();
            let body = http_response.bytes().await.map_err(map_body_error)?;
            finalize_response(status.as_u16(), &response_headers, &body, decode, response)
        }
    }
}

/// Sync reqwest::blocking-based REST transport implementing [`SyncUnaryTransport`].
#[derive(Clone)]
pub struct SyncRestTransport {
    base_url: String,
    auth: Arc<dyn Auth>,
    client: reqwest::blocking::Client,
}

impl SyncRestTransport {
    /// Creates a sync REST transport rooted at `base_url`.
    pub fn new(base_url: impl Into<String>, auth: Arc<dyn Auth>) -> Self {
        Self {
            base_url: base_url.into(),
            auth,
            client: reqwest::blocking::Client::builder()
                .use_rustls_tls()
                .build()
                .expect("reqwest blocking client"),
        }
    }

    /// Applies a per-request timeout to the underlying HTTP client.
    pub fn with_timeout(mut self, timeout: Duration) -> Self {
        self.client = reqwest::blocking::Client::builder()
            .use_rustls_tls()
            .timeout(timeout)
            .build()
            .expect("reqwest blocking client");
        self
    }
}

impl SyncUnaryTransport for SyncRestTransport {
    fn unary<Req, Resp>(
        &self,
        method: &Method,
        request: &Req,
        response: &mut Resp,
    ) -> Result<(), GestaltError>
    where
        Req: Message + Clone + Send + Sync + 'static,
        Resp: Message + Default + Send + 'static,
    {
        let decode = method.decode_response_json.ok_or_else(|| {
            GestaltError::new(
                gestalt_error_code::INVALID_ARGUMENT,
                format!("method {} has no REST decoder", method.full_method),
            )
        })?;
        let prepared =
            build_rest_request(method, &request.encode_to_vec(), &self.base_url, &self.auth)?;
        let mut builder = self
            .client
            .request(prepared.method, prepared.url)
            .headers(prepared.headers);
        if let Some(body) = prepared.body {
            builder = builder.json(&body);
        }
        let http_response = builder.send().map_err(map_send_error)?;
        let status = http_response.status();
        let response_headers = http_response.headers().clone();
        let body = http_response.bytes().map_err(map_body_error)?;
        finalize_response(status.as_u16(), &response_headers, &body, decode, response)
    }
}
